xla/hlo/analysis/while_loop_analysis_test.cc

```diff
@@ -425,5 +425,159 @@ TEST_F(WhileLoopAnalysisTest, AIVNoChain) {
+    }
+  )";
+
+  TF_ASSERT_OK_AND_ASSIGN(auto module, ParseAndReturnVerifiedModule(hlo));
+  const HloInstruction* while_op =
+      module->entry_computation()->root_instruction();
+  EXPECT_EQ(ComputeWhileLoopTripCount(while_op), std::nullopt);
+}
+
+TEST_F(WhileLoopAnalysisTest, FusedUpdateOp) {
+  const char* hlo = R"(
+  HloModule test, replica_count=2
+  add {
+    param.3 = s32[] parameter(0)
+    param.4 = s32[] parameter(1)
+    ROOT add = add(param.3, param.4)
+  }
+  body {
+    param.0 = (s32[], s32[]) parameter(0)
+    p0.1 = s32[] get-tuple-element(param.0), index=0
+    p1.1 = s32[] get-tuple-element(param.0), index=1
+    c1 = s32[] constant(1)
+    update = s32[] fusion(p0.1, c1), kind=kInput, calls=add
+    ROOT tuple = (s32[], s32[]) tuple(update, p1.1)
+  }
+  condition {
+    param.2 = (s32[], s32[]) parameter(0)
+    p0.2 = s32[] get-tuple-element(param.2), index=0
+    c4 = s32[] constant(4)
+    ROOT compare = pred[] compare(p0.2, c4), direction=LT
+  }
+  ENTRY entry {
+    c0 = s32[] constant(0)
+    data = s32[] parameter(0)
+    tuple = (s32[], s32[]) tuple(c0, data)
+    ROOT while = (s32[], s32[]) while(tuple), body=body, condition=condition
+  }
+  )";
+  TF_ASSERT_OK_AND_ASSIGN(auto module, ParseAndReturnVerifiedModule(hlo));
+  HloInstruction* while_op = module->entry_computation()->root_instruction();
+  auto trip_count = ComputeWhileLoopTripCount(while_op);
+  ASSERT_NE(trip_count, std::nullopt);
+  EXPECT_EQ(trip_count, 4);
+}
+
+TEST_F(WhileLoopAnalysisTest, NonScalarConditionOp) {
+  const char* hlo = R"(
+    HloModule test, replica_count=2
+    add {
+      param.3 = s32[] parameter(0)
+      param.4 = s32[] parameter(1)
+      ROOT add = add(param.3, param.4)
+    }
+    body {
+      param.0 = (s32[], s32[]) parameter(0)
+      p0.1 = s32[] get-tuple-element(param.0), index=0
+      p1.1 = s32[] get-tuple-element(param.0), index=1
+      c1 = s32[] constant(1)
+      update = s32[] add(p0.1, c1)
+      ROOT tuple = (s32[], s32[]) tuple(update, p1.1)
+    }
+    fused_computation {
+      param.5 = s32[] parameter(0)
+      param.6 = s32[] parameter(1)
+      all-reduce.1 = s32[] all-reduce(param.5), replica_groups={{0,1}}, to_apply=add
+      ROOT compare = pred[] compare(all-reduce.1, param.6), direction=LT
+    }
+    condition {
+      param.2 = (s32[], s32[]) parameter(0)
+      p0.2 = s32[] get-tuple-element(param.2), index=0
+      c4 = s32[] constant(4)
+      ROOT compare = pred[] call(p0.2, c4), to_apply=fused_computation
+    }
+    ENTRY entry {
+      c0 = s32[] constant(0)
+      data = s32[] parameter(0)
+      tuple = (s32[], s32[]) tuple(c0, data)
+      ROOT while = (s32[], s32[]) while(tuple), body=body, condition=condition
+    }
+  )";
+
+  TF_ASSERT_OK_AND_ASSIGN(auto module, ParseAndReturnVerifiedModule(hlo));
+  const HloInstruction* while_op =
+      module->entry_computation()->root_instruction();
+  EXPECT_EQ(ComputeWhileLoopTripCount(while_op), std::nullopt);
+}
+
```

Comment:
Can you please add a test case showing what kind of things I want to be guarded against?

E.g. a condition computation like this:

```
or {
  x = pred[] parameter(0)
  y = pred[] parameter(1)
  ROOT res = pred[] or(x, y)
}
cond {
  param.0 = (pred[100000000]) parameter(0)
  gte = pred[100000000] get-tuple-element(param.0), index=0
  false = pred[] constant(false)
  ROOT any = pred[] reduce(gte, false), dimensions={0}, to_apply=or
}
```
(needs some tweaking to the parameter tuple to have extra entries and some matching body computation, but I hope you get the idea - for example that indvar tuple entry could be the result of a compare op in the body computation).

Reply:
Added the test.